File: app/src/main/java/com/duzhaokun123/bilibilihd/ui/widget/BiliPlayerViewWrapperView.java
package com.duzhaokun123.bilibilihd.ui.widget;

import android.content.Context;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.PopupMenu;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;

import com.duzhaokun123.bilibilihd.R;
import com.duzhaokun123.bilibilihd.databinding.LayoutBiliPlayerViewWapperViewBinding;
import com.duzhaokun123.bilibilihd.utils.GlideUtil;
import com.duzhaokun123.bilibilihd.utils.TipUtil;
import com.google.android.exoplayer2.ExoPlaybackException;
import com.google.android.exoplayer2.Player;
import com.google.android.exoplayer2.SimpleExoPlayer;
import com.google.android.exoplayer2.source.MediaSource;
import com.google.android.exoplayer2.ui.PlayerControlView;

import java.util.List;

import kotlin.Pair;
import master.flame.danmaku.danmaku.parser.BaseDanmakuParser;

public class BiliPlayerViewWrapperView extends FrameLayout {
    private LayoutBiliPlayerViewWapperViewBinding baseBind;

    private OnPlayingStatusChangeListener onPlayingStatusChangeListener;
    private OnPlayerErrorListener onPlayerErrorListener;
    private VideoMediaSourceAdapter videoMediaSourceAdapter;

    private SimpleExoPlayer player;

    private int qualityId = 0;

    public BiliPlayerViewWrapperView(@NonNull Context context) {
        this(context, null);
    }

    public BiliPlayerViewWrapperView(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public BiliPlayerViewWrapperView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        if (isInEditMode()) {
            LayoutInflater.from(context).inflate(R.layout.layout_bili_player_view_wapper_view, this);
            return;
        }

        baseBind = DataBindingUtil.inflate(LayoutInflater.from(context), R.layout.layout_bili_player_view_wapper_view, this, true);
        player = new SimpleExoPlayer.Builder(context).build();
        baseBind.bpv.setPlayer(player);
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        if (isInEditMode()) {
            return;
        }
        player.addListener(new Player.EventListener() {
            @Override
            public void onSeekProcessed() {
                baseBind.bpv.danmakuSeekTo(player.getContentPosition());
            }

            @Override
            public void onIsPlayingChanged(boolean isPlaying) {
                if (isPlaying) {
                    baseBind.bpv.danmakuResume();
                    baseBind.bpv.setPbExoBufferingVisibility(View.INVISIBLE);
                    if (onPlayingStatusChangeListener != null) {
                        onPlayingStatusChangeListener.onPlayingStatusChange(PlayingStatus.PLAYING);
                    }
                } else {
                    baseBind.bpv.danmakuPause();
                    if (onPlayingStatusChangeListener != null) {
                        onPlayingStatusChangeListener.onPlayingStatusChange(PlayingStatus.PAUSED);
                    }
                    long contentPosition = player.getContentPosition();
                    long contentDuration = player.getContentDuration();
                    long contentBufferedPosition = player.getContentBufferedPosition();
                    if (contentBufferedPosition - contentPosition <= 1000 && contentDuration - contentPosition > 100) {
                        baseBind.bpv.setPbExoBufferingVisibility(View.VISIBLE);
                        if (onPlayingStatusChangeListener != null) {
                            onPlayingStatusChangeListener.onPlayingStatusChange(PlayingStatus.LOADING);
                        }
                    } else {
                        baseBind.bpv.showController();
                        if (onPlayingStatusChangeListener != null) {
                            onPlayingStatusChangeListener.onPlayingStatusChange(PlayingStatus.ENDED);
                        }
                    }
                }
            }

            @Override
            public void onPlayerError(@NonNull ExoPlaybackException error) {
                if (onPlayingStatusChangeListener != null) {
                    onPlayingStatusChangeListener.onPlayingStatusChange(PlayingStatus.ERROR);
                }
                if (onPlayerErrorListener != null) {
                    onPlayerErrorListener.onPlayerError(error);
                }
            }
        });
        baseBind.ivCover.setOnClickListener(view -> {
            setCover(null);
            resume();
        });
        baseBind.bpv.setDanmakuLoadListener(e -> {
            if (e != null) {
                TipUtil.showTip(getContext(), e.getMessage());
            }
        });
        baseBind.bpv.getBtnQuality().setOnClickListener(view -> {
            if (videoMediaSourceAdapter == null) {
                return;
            }

            PopupMenu popupMenu = new PopupMenu(getContext(), baseBind.bpv.getBtnQuality());
            Menu menu = popupMenu.getMenu();
            for (int i = 0; i < videoMediaSourceAdapter.getCount(); i++) {
                String name = videoMediaSourceAdapter.getName(i);
                menu.add(0, i, i, name != null ? name : String.valueOf(i));
            }
            popupMenu.setOnMenuItemClickListener(menuItem -> {
                changeQuality(menuItem.getItemId());
                return true;
            });
            popupMenu.show();
        });
        baseBind.bpv.setOnClickListener(new OnClickListener() {
            private long lastClickTime = 0;

            @Override
            public void onClick(View v) {
                long thisClickTime = System.currentTimeMillis();
                if (thisClickTime - lastClickTime < 400) {
                    if (player.getPlayWhenReady()) {
                        pause();
                    } else {
                        resume();
                    }
                }
                lastClickTime = thisClickTime;
            }
        });
    }

    public void release() {
        player.release();
        baseBind.bpv.release();
    }

    public void setCover(String url) {
        if (url != null) {
            baseBind.ivCover.setVisibility(VISIBLE);
            GlideUtil.loadUrlInto(getContext(), url, baseBind.ivCover, false);
        } else {
            baseBind.ivCover.setVisibility(GONE);
            baseBind.ivCover.setImageDrawable(null);
        }
    }

    public void pause() {
        player.setPlayWhenReady(false);
        baseBind.bpv.danmakuPause();
    }

    public void resume() {
        player.setPlayWhenReady(true);
        baseBind.bpv.danmakuResume();
    }

    public OnFullscreenClickListener getOnFullscreenClickListener() {
        return baseBind.bpv.getOnFullscreenClickListener();
    }

    public void setOnFullscreenClickListener(OnFullscreenClickListener onFullscreenClickListener) {
        baseBind.bpv.setOnFullscreenClickListener(onFullscreenClickListener);
    }

    public OnPlayingStatusChangeListener getOnPlayingStatusChangeListener() {
        return onPlayingStatusChangeListener;
    }

    public void setOnPlayingStatusChangeListener(OnPlayingStatusChangeListener onPlayingStatusChangeListener) {
        this.onPlayingStatusChangeListener = onPlayingStatusChangeListener;
    }

    public void setControllerVisibilityListener(PlayerControlView.VisibilityListener visibilityListener) {
        baseBind.bpv.setControllerVisibilityListener(visibilityListener);
    }

    public BiliPlayerView getBiliPlayerView() {
        return baseBind.bpv;
    }

    public void loadDanmakuByAidCid(long aid, long cid, int durationS) {
        baseBind.bpv.loadDanmakuByAidCid(aid, cid, durationS);
    }

    public void loadDanmakuByBiliDanmakuParser(BaseDanmakuParser danmakuParser) {
        baseBind.bpv.loadDanmakuByBiliDanmakuParser(danmakuParser);
    }

    public void setOnPlayerErrorListener(OnPlayerErrorListener onPlayerErrorListener) {
        this.onPlayerErrorListener = onPlayerErrorListener;
    }

    public VideoMediaSourceAdapter getVideoMediaSourceAdapter() {
        return videoMediaSourceAdapter;
    }

    public void setVideoMediaSourceAdapter(VideoMediaSourceAdapter videoMediaSourceAdapter) {
        this.videoMediaSourceAdapter = videoMediaSourceAdapter;
        player.seekTo(0);
        changeQuality(videoMediaSourceAdapter.getDefaultIndex());
    }

    private void changeQuality(int index) {
        if (videoMediaSourceAdapter == null) {
            baseBind.bpv.getBtnQuality().setVisibility(GONE);
        } else {
            int count = videoMediaSourceAdapter.getCount();
            if (count > index) {
                int qualityIdTemp = videoMediaSourceAdapter.getId(index);
                baseBind.bpv.getBtnQuality().setVisibility(VISIBLE);
                List<Pair<String, MediaSource>> mediaSources = videoMediaSourceAdapter.getMediaSources(qualityIdTemp);
                if (mediaSources != null) {
                    updateMediaSource(mediaSources.get(0).getSecond());
                    baseBind.bpv.getBtnQuality().setText(videoMediaSourceAdapter.getName(index));
                    baseBind.bpv.getBtnLine().setText(mediaSources.get(0).getFirst());
                    qualityId = qualityIdTemp;
                    baseBind.bpv.getBtnLine().setOnClickListener(
                            new OnClickListener() {
                                private int order = 0;

                                @Override
                                public void onClick(View v) {
                                    PopupMenu popupMenu = new PopupMenu(getContext(), v);
                                    for (int i = 0; i < mediaSources.size(); i++) {
                                        popupMenu.getMenu().add(0, i, i, mediaSources.get(i).getFirst());
                                    }
                                    popupMenu.setOnMenuItemClickListener(item -> {
                                                if (order != item.getOrder()) {
                                                    order = item.getOrder();
                                                    updateMediaSource(mediaSources.get(order).getSecond());
                                                    baseBind.bpv.getBtnLine().setText(mediaSources.get(order).getFirst());

                                                }
                                                return true;
                                            }
                                    );
                                    popupMenu.show();
                                }


                            }
                    );
                }
            }
        }
    }


    public SimpleExoPlayer getPlayer() {
        return player;
    }

    public boolean isPlaying() {
        return player.isPlaying();
    }

    public void clickIbFullscreen() {
        baseBind.bpv.clickIbFullscreen();
    }

    public int getQualityId() {
        return qualityId;
    }

    public void setQualityId(int qualityId) {
        this.qualityId = qualityId;
    }

    public void syncDanmakuProgress() {
        baseBind.bpv.getDanmakuView().seekTo(player.getContentPosition());
    }

    public void setLive(boolean isLive) {
        baseBind.bpv.setLive(isLive);
    }

    public void setOnDanmakuSendClickListener(OnClickListener listener) {
        baseBind.bpv.setDanmakuSendClickListener(listener);
    }

    private void updateMediaSource(MediaSource mediaSource) {
        long playedTime = player.getCurrentPosition();
        player.prepare(mediaSource);
        player.seekTo(playedTime);
    }

    public interface OnFullscreenClickListener {
        void onClick(boolean isFullscreen);
    }

    public interface OnPlayerErrorListener {
        void onPlayerError(ExoPlaybackException error);
    }

    public interface OnPlayingStatusChangeListener {
        void onPlayingStatusChange(@NonNull PlayingStatus playingStatus);
    }

    public interface VideoMediaSourceAdapter {
        @Nullable
        List<Pair<String, MediaSource>> getMediaSources(int id);

        int getCount();

        @Nullable
        String getName(int index);

        int getDefaultIndex();

        int getId(int index);
    }

    public enum PlayingStatus {
        PLAYING,
        PAUSED,
        LOADING,
        ENDED,
        ERROR
    }
}

File: app/src/main/java/com/duzhaokun123/bilibilihd/ui/widget/BiliPlayerViewWrapperView.kt
package com.duzhaokun123.bilibilihd.ui.widget

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.widget.FrameLayout
import android.widget.PopupMenu
import androidx.databinding.DataBindingUtil
import com.duzhaokun123.bilibilihd.R
import com.duzhaokun123.bilibilihd.databinding.LayoutBiliPlayerViewWapperViewBinding
import com.duzhaokun123.bilibilihd.utils.GlideUtil
import com.duzhaokun123.bilibilihd.utils.TipUtil
import com.duzhaokun123.danmakuview.interfaces.DanmakuParser
import com.duzhaokun123.danmakuview.model.Danmakus
import com.google.android.exoplayer2.ExoPlaybackException
import com.google.android.exoplayer2.Player
import com.google.android.exoplayer2.SimpleExoPlayer
import com.google.android.exoplayer2.source.MediaSource
import com.google.android.exoplayer2.ui.PlayerControlView

class BiliPlayerViewWrapperView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : FrameLayout(context, attrs, defStyleAttr) {
    private lateinit var baseBind: LayoutBiliPlayerViewWapperViewBinding

    var onPlayingStatusChangeListener: ((playState: PlayingStatus) -> Unit)? = null

    var onPlayerErrorListener: ((error: ExoPlaybackException) -> Unit)? = null

    var videoMediaSourceAdapter: VideoMediaSourceAdapter? = null
        set(value) {
            field = value
            player.seekTo(0)
            if (value != null) {
                baseBind.bpv.btnLine.visibility = VISIBLE
                baseBind.bpv.btnQuality.visibility = VISIBLE
                changeQuality(value.defaultIndex)
            } else {
                baseBind.bpv.btnLine.visibility = GONE
                baseBind.bpv.btnQuality.visibility = GONE
            }
        }

    lateinit var player: SimpleExoPlayer
        private set

    var qualityId = 0

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        if (isInEditMode) {
            return
        }
        player.addListener(object : Player.EventListener {
            override fun onSeekProcessed() {
                baseBind.bpv.danmakuSeekTo(player.contentPosition)
            }

            override fun onIsPlayingChanged(isPlaying: Boolean) {
                if (isPlaying) {
                    baseBind.bpv.danmakuResume()
                    baseBind.bpv.setPbExoBufferingVisibility(INVISIBLE)
                    if (onPlayingStatusChangeListener != null) {
                        onPlayingStatusChangeListener!!(PlayingStatus.PLAYING)
                    }
                } else {
                    baseBind.bpv.danmakuPause()
                    if (onPlayingStatusChangeListener != null) {
                        onPlayingStatusChangeListener!!(PlayingStatus.PAUSED)
                    }
                    val contentPosition = player.contentPosition
                    val contentDuration = player.contentDuration
                    val contentBufferedPosition = player.contentBufferedPosition
                    if (contentBufferedPosition - contentPosition <= 1000 && contentDuration - contentPosition > 100) {
                        baseBind.bpv.setPbExoBufferingVisibility(VISIBLE)
                        if (onPlayingStatusChangeListener != null) {
                            onPlayingStatusChangeListener!!(PlayingStatus.LOADING)
                        }
                    } else {
                        baseBind.bpv.showController()
                        if (onPlayingStatusChangeListener != null) {
                            onPlayingStatusChangeListener!!(PlayingStatus.ENDED)
                        }
                    }
                }
            }

            override fun onPlayerError(error: ExoPlaybackException) {
                if (onPlayingStatusChangeListener != null) {
                    onPlayingStatusChangeListener!!(PlayingStatus.ERROR)
                }
                if (onPlayerErrorListener != null) {
                    onPlayerErrorListener!!(error)
                }
            }
        })
        baseBind.ivCover.setOnClickListener {
            setCover(null)
            resume()
        }
        baseBind.bpv.setDanmakuLoadListener { e: Exception? ->
            if (e != null) {
                TipUtil.showTip(context, e.message)
            }
        }
        baseBind.bpv.btnQuality.setOnClickListener {
            if (videoMediaSourceAdapter == null) {
                return@setOnClickListener
            }
            val popupMenu = PopupMenu(context, baseBind.bpv.btnQuality)
            val menu = popupMenu.menu
            for (i in 0 until videoMediaSourceAdapter!!.count) {
                val name = videoMediaSourceAdapter!!.getName(i)
                menu.add(0, i, i, name ?: i.toString())
            }
            popupMenu.setOnMenuItemClickListener { menuItem: MenuItem ->
                changeQuality(menuItem.itemId)
                true
            }
            popupMenu.show()
        }
        baseBind.bpv.setOnClickListener(object : OnClickListener {
            private var lastClickTime: Long = 0
            override fun onClick(v: View) {
                val thisClickTime = System.currentTimeMillis()
                if (thisClickTime - lastClickTime < 400) {
                    if (player.playWhenReady) {
                        pause()
                    } else {
                        resume()
                    }
                }
                lastClickTime = thisClickTime
            }
        })
    }

    fun release() {
        player.release()
        baseBind.bpv.release()
    }

    fun setCover(url: String?) {
        if (url != null) {
            baseBind.ivCover.visibility = VISIBLE
            GlideUtil.loadUrlInto(context, url, baseBind.ivCover, false)
        } else {
            baseBind.ivCover.visibility = GONE
            baseBind.ivCover.setImageDrawable(null)
        }
    }

    fun pause() {
        player.playWhenReady = false
        baseBind.bpv.danmakuPause()
    }

    fun resume() {
        player.playWhenReady = true
        baseBind.bpv.danmakuResume()
    }

    var onFullscreenClickListener: OnFullscreenClickListener?
        get() = baseBind.bpv.onFullscreenClickListener
        set(onFullscreenClickListener) {
            baseBind.bpv.onFullscreenClickListener = onFullscreenClickListener
        }

    fun setControllerVisibilityListener(visibilityListener: PlayerControlView.VisibilityListener?) {
        baseBind.bpv.setControllerVisibilityListener(visibilityListener)
    }

    val biliPlayerView: BiliPlayerView
        get() = baseBind.bpv

    fun loadDanmakuByAidCid(aid: Long, cid: Long, durationS: Int) {
        baseBind.bpv.loadDanmakuByAidCid(aid, cid, durationS)
    }

    fun loadDanmakuByBiliDanmakuParser(danmakuParser: DanmakuParser, onEnd: ((danmakus: Danmakus) -> Unit)? =null) {
        baseBind.bpv.danmakuView.parse(danmakuParser, onEnd)
    }

    private fun changeQuality(index: Int) {
        if (videoMediaSourceAdapter != null) {
            val count = videoMediaSourceAdapter!!.count
            if (count > index) {
                val qualityIdTemp = videoMediaSourceAdapter!!.getId(index)
                baseBind.bpv.btnQuality.visibility = VISIBLE
                val mediaSources = videoMediaSourceAdapter!!.getMediaSources(qualityIdTemp)
                if (mediaSources != null) {
                    updateMediaSource(mediaSources[0].second)
                    baseBind.bpv.btnQuality.text = videoMediaSourceAdapter!!.getName(index)
                    baseBind.bpv.btnLine.text = mediaSources[0].first
                    qualityId = qualityIdTemp
                    baseBind.bpv.btnLine.setOnClickListener(
                            object : OnClickListener {
                                private var order = 0
                                override fun onClick(v: View) {
                                    val popupMenu = PopupMenu(context, v)
                                    for (i in mediaSources.indices) {
                                        popupMenu.menu.add(0, i, i, mediaSources[i].first)
                                    }
                                    popupMenu.setOnMenuItemClickListener { item: MenuItem ->
                                        if (order != item.order) {
                                            order = item.order
                                            updateMediaSource(mediaSources[order].second)
                                            baseBind.bpv.btnLine.text = mediaSources[order].first
                                        }
                                        true
                                    }
                                    popupMenu.show()
                                }
                            }
                    )
                }
            }
        }
    }

    val isPlaying: Boolean
        get() = player.isPlaying

    fun clickIbFullscreen() {
        baseBind.bpv.clickIbFullscreen()
    }

    fun syncDanmakuProgress() {
        baseBind.bpv.danmakuView.seekTo(player.contentPosition)
    }

    fun setLive(isLive: Boolean) {
        baseBind.bpv.setLive(isLive)
    }

    fun setOnDanmakuSendClickListener(listener: OnClickListener?) {
        baseBind.bpv.setDanmakuSendClickListener(listener)
    }

    private fun updateMediaSource(mediaSource: MediaSource) {
        val playedTime = player.currentPosition
        player.prepare(mediaSource)
        player.seekTo(playedTime)
    }

    interface OnFullscreenClickListener {
        fun onClick(isFullscreen: Boolean)
    }

    interface VideoMediaSourceAdapter {
        fun getMediaSources(id: Int): List<Pair<String, MediaSource>>?
        val count: Int
        fun getName(index: Int): String?
        val defaultIndex: Int
        fun getId(index: Int): Int
    }

    enum class PlayingStatus {
        PLAYING, PAUSED, LOADING, ENDED, ERROR
    }

    init {
        if (isInEditMode) {
            LayoutInflater.from(context).inflate(R.layout.layout_bili_player_view_wapper_view, this)

        } else {
            baseBind = DataBindingUtil.inflate(LayoutInflater.from(context), R.layout.layout_bili_player_view_wapper_view, this, true)
            player = SimpleExoPlayer.Builder(context).build()
            baseBind.bpv.player = player
        }
    }
}